containers-tests/tests/map-properties.hs

```diff
@@ -1365,12 +1371,46 @@ prop_splitModel n ys = length ys > 0 ==>
   in  toAscList l == sort [(k, v) | (k,v) <- xs, k < n] &&
       toAscList r == sort [(k, v) | (k,v) <- xs, k > n]
 
+prop_fold :: [(Int, Int)] -> Property
+prop_fold ys = length ys > 0 ==>
+  let xs = List.nubBy ((==) `on` fst) ys
+      m  = fromList xs
+  in  getSum (Foldable.fold (map Sum m)) == sum (List.map snd xs) &&
+#if MIN_VERSION_base(4,10,0)
+      -- This seems to fit the pattern of the other test cases, but isn't valid here:
+      --getSum (bifold (mapKeys (const mempty) (map Sum m))) == sum (List.map fst xs) &&
+      -- since it would map all of the keys to the same thing, thus removing all but one element.
+      getSum (bifold (mapKeys Sum (map (const mempty) m))) == sum (List.map fst xs) &&
+      bifold (mapKeys (:[]) (map (:[]) m)) == concatMap (\(k,v) -> [k,v]) (List.sort xs) &&
```

Comment:
I don't understand what any of this crud is doing. I think we should either use the list monoid exclusively or, if that is too expensive, the sequence monoid. I don't care for the nubbing or the use of `fromList`.  Take a `Map` as an argument (which gives us lots of ways to balance arbitrarily) and use `toList`.

Reply:
But `toList` is implemented in terms of `foldrWithKey`, so wouldn't doing that make the tests basically worthless, since `foldrWithKey` is one of the things we're supposed to be testing?